crates/rpc/rpc/src/trace.rs

```diff
@@ -549,6 +579,41 @@ where
         }
         traces
     }
+
+    /// Returns all storage slots accessed during transaction execution along with their access
+    /// counts.
+    pub async fn trace_block_storage_access(
+        &self,
+        block_id: BlockId,
+    ) -> Result<Option<BlockStorageAccess>, Eth::Error> {
+        let res = self
+            .eth_api()
+            .trace_block_inspector(
+                block_id,
+                None,
+                StorageInspector::default,
+                move |tx_info, inspector, _res, _, _| {
+                    let trace = TransactionStorageAccess {
+                        transaction_hash: tx_info.hash.expect("tx hash is set"),
+                        storage_access: inspector.accessed_slots().clone(),
+                        unique_loads: inspector.unique_loads(),
+                        warm_loads: inspector.warm_loads(),
+                    };
+                    Ok(trace)
+                },
+            )
+            .await?;
+
+        let Some(transactions) = res else { return Ok(None) };
+
+        let Some(block) = self.eth_api().recovered_block(block_id).await? else { return Ok(None) };
```

Comment:
instead of doing this here just to load the hash, we can force convert the blockid to blockhash

https://github.com/paradigmxyz/reth/blob/f24570844545b55f2293a3a3c475133a3f738b68/crates/rpc/rpc-eth-api/src/helpers/block.rs#L241-L243